doc/dependency-resolution.md

```diff
@@ -0,0 +1,319 @@
+receive one of these interactions for a given content_id, we should explicitly
+check its dependencies and send updates downstream. Because the rules for how
+dependencies are tracked is captured in the content schemas, it is proposed that
+Publishing API be granted visibility over the content schemas in production so
+that it can apply these rules when detecting changes. We had already planned to
+do this for the purpose of validating content against the schemas in the
+Publishing API.
+
+Once it has visibility over the dependency rules for each of the groups, it
+should run a query against the database to detect content items that need to be
+updated as a result of the incoming request. It will do this using Postgres's
+WITH RECURSIVE functionality. This was explored in the following branch to
+demonstrate that it is feasible to detect dependent nodes in this way:
+
+[https://github.com/alphagov/publishing-api/tree/dependency-transitive-closure](https://github.com/alphagov/publishing-api/tree/dependency-transitive-closure)
+
+We should make it so that this query only follows edges in the dependency graph
+for groups which contain any of the fields that have changed on the content
+item. We can determine which fields have changed using Rails's
+[ActiveModel::Dirty](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html)
+methods. The branch linked to above demonstrates that it is possible to restrict
+edges followed by a link_type (which is another name for 'group'). The recursive
+behaviour of this query depends on the boolean configuration setting for that
+group and should be factored into the query.
+
+All of this behaviour should be completed out-of-band from the publishing
+application's request cycle with the Publishing API. It is proposed that we
+isolate all of this dependency resolution work into a separate Sidekiq worker
+that is responsible for determining the set of content items that need to be
+updated and for actually presenting and sending those updates downstream. When
+the Publishing API receives any interaction that can affect dependencies, it
+need only place this content item on a 'dirty' queue for processing by the
+dependency resolution worker.
+
+If required, we can separate the calculation of dependencies from sending the
+updates downstream into separate workers/queues. The important thing is that
+both of those tasks happen out-of-band to the request cycle.
+
+Note: Bidirectional dependencies is documented separately later on.
+
+- Link expansion
+
+It is proposed that we move link expansion into the publishing api and send the
+fully-pressed content item downstream to the content store. For each of the link
+groups that appear on the content item, it will need to look up the rules for
+those groups from the schemas and expand those fields from the dependent content
+items.
+
+We also need to consider a new data structure to represent recursively expanded
```

Comment:
Added a rough example.

We're going to design this structure relatively early in the development of dependency resolution, so will be able to provide a more detailed description of this once we start development.